source/server/connection_handler_impl.cc

```diff
@@ -458,6 +458,9 @@ void ConnectionHandlerImpl::ActiveTcpListener::resumeListening() {
 
 void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
     Network::ConnectionSocketPtr&& socket, std::unique_ptr<StreamInfo::StreamInfo> stream_info) {
+  // Update local address in case it is restored by a listener filter
```

Comment:
Can you add the TODO? This is good since we known original_dst could change setDownstreamLocalAddress but we don't know if other crazy listener filter could do (proxy protocol?)

Reply:
Without exposing stream info to listener filters, we just have to fix this case by case. Is there something else that is changed by listener filters?

Reply:
TransportProtocol applicationProtocol. It depends on the overlap of stream info and socket info.
Also newConnection() is not the only termination of the socket. For a closed socket the small discrepancy is not ideal but should be fine for waiting for the further solution.

Reply:
I'm having some trouble understanding the issue being fixed in this PR.  Should original_dst be calling setDownstreamLocalAddress?  I'm guessing that the answer is no.

I'm concerned that this change breaks other intended behavior, but I can see the counter argument involving no existing tests breaking.  That said, a lot of functionality is covered by small tests, which wouldn't detect breakages due to changes like this one which involve interaction between multiple extensions.

Reply:
The history is that this is a regression. Prior to https://github.com/envoyproxy/envoy/issues/13554, stream info was instantiated after listener filters, so the value of downstream local address is the modified value. After that PR, stream info is instantiated before listener filters, so the value has changed. This is a fix to restore the original behavior.

The real issue is lack of any integration testing for original_dst. I think we would have caught the regression sooner if we had any test that exercised original_dst and logging. I don't know why it's like that, maybe it was difficult to set up iptables in a test.

Reply:
Is socket->localAdress() the actual address coming from the socket or a modified value set by orig dst filter?

Basically, I'm having trouble understanding new behavior based on the comment added here.

Reply:
It's the modified value AFAICT.